### PR TITLE
app: added high_registration_failures_rate check

### PR DESCRIPTION
**Pull request by pinebit**
Added `high_registration_failures_rate` app health check.
It triggers a warning if the registration failures rate goes above 70% for two consecutive epochs.

category: feature
ticket: #2760

### CODE REVIEW

File: app/health/checks.go [hunk 1]
```diff
@@ -123,6 +125,19 @@ var checks = []check{
 			return increase > 0, nil
 		},
 	},
+	{
+		Name:        "high_registration_failures_rate",
+		Description: "High rate of failed registrations. Please check the logs for more details.",
```
Comment (xenowits) [inline app/health/checks.go:130]:
```suggestion
		Description: "High rate of failed validator registrations. Please check the logs for more details.",
```

File: core/bcast/errors_rate.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+package bcast
+
+import (
+	"sync/atomic"
+
+	"github.com/obolnetwork/charon/core"
+)
+
+const (
+	// RegistrationErrorsRateThreshold defines the threshold for registration errors rate.
```
Comment (gsora) [inline core/bcast/errors_rate.go:12]:
```suggestion
	// BuilderRegistrationErrorsRateThreshold defines the threshold for builder registration errors rate.
```

Comment (gsora) [inline core/bcast/errors_rate.go:12]:
```suggestion
	// BuilderRegistrationErrorsRateThreshold defines the threshold for builder registration errors rate.
```

File: core/bcast/errors_rate.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	RegistrationErrorsRateThreshold int32 = 70
```
Comment (gsora) [inline core/bcast/errors_rate.go:13]:
```suggestion
	BuilderRegistrationErrorsRateThreshold int32 = 70
```

File: core/bcast/errors_rate.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+)
+
+// errorsRate object tracks registration errors rate within the last two epochs and updates metrics.
+type errorsRate struct {
```
Comment (gsora) [inline core/bcast/errors_rate.go:17]:
```suggestion
type builderRegistrationErrorsRate struct {
```

Not the best name ever, but errorsRate seems a little too generic.

File: app/health/checks.go [hunk 5]
```diff
@@ -123,6 +125,19 @@ var checks = []check{
 			return increase > 0, nil
 		},
 	},
+	{
+		Name:        "high_registration_failures_rate",
+		Description: "High rate of failed validator registrations. Please check the logs for more details.",
+		Severity:    severityWarning,
+		Func: func(q query, m Metadata) (bool, error) {
+			max, err := q("core_bcast_recast_errors_rate", sumLabels(), gaugeMax)
```
Comment (dB2510) [inline app/health/checks.go:133]: why do we need a different metric for this? We can calculate the rate from the existing metric: `core_bcast_recast_errors_total`

File: core/bcast/errors_rate.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+	recastTotal.WithLabelValues(regSourcePregen).Add(float64(r.pregenState.total.Load()))
+	recastErrors.WithLabelValues(regSourcePregen).Add(float64(r.pregenState.errors.Load()))
+	recastTotal.WithLabelValues(regSourceDownstream).Add(float64(r.downstreamState.total.Load()))
+	recastErrors.WithLabelValues(regSourceDownstream).Add(float64(r.downstreamState.errors.Load()))
+
+	currentPregenRate := calculateRate(&r.pregenState)
+	currentDownstreamRate := calculateRate(&r.downstreamState)
+
+	// We want to be tolerant of the request failing in a single epoch, and warn when it happens for two consecutive epochs.
+	if r.pregenState.prevRate.Load() >= RegistrationErrorsRateThreshold && currentPregenRate >= RegistrationErrorsRateThreshold {
```
Comment (gsora) [inline core/bcast/errors_rate.go:66]: The way it’s implemented, `recastErrorsRate` act as a boolean flag right?

E.g. if in the previous epoch we failed 10% of the requests, and 75% in the current epoch, we won’t trigger the warning.

File: app/health/checks.go (hunk 5 above)
Comment (pinebit) [reply app/health/checks.go:133]: To calculate percentage you need to know two parameters: the number of errors and the total number of samples. We can be querying two metrics, but we also needed to address the new requirement about two consecutive epochs.
I don't see an easy way to provide epics information here to be able to compare the last twos.

File: core/bcast/errors_rate.go (hunk 6 above)
Comment (pinebit) [reply core/bcast/errors_rate.go:66]: Taken offline for internal discussion.

**Comment by pinebit** [comment]
Closing in favor of simpler design.

**Pull request closed without merging**